### DISCUSSION TITLE
Lesson 6: Extend your test to better understand the mystery bug: AttributeError: 'NoneType'

### DISCUSSION THREAD

**Discussion by n4n0b1t3** [smartcontractkit/full-blockchain-solidity-course-py#1045] [Ideas]
If you wrote your test in Lesson 6 without first checking the repository to see what the final version looks like, you will have come across a strange error:
`E       AttributeError: 'NoneType' object has no attribute '_with_attr'`

I cannot tell from the error message what happened here. Only a look at the repository shows that 100 Wei were added to the entrance_fee as a buffer so that the conversion does not end up below the admission threshold. 

I, therefore, suggest extending the test as follows, so that it then becomes quite clear what went wrong. This is especially important if you don't have a working template but write your own code.

```Python
from scripts.helpers import get_account
from scripts.deploy import deploy_fund_me

def test_can_fund_and_withdraw():
    account = get_account()
    fund_me = deploy_fund_me()
    entrance_fee = fund_me.getEntranceFee()
    # entrance_fee = fund_me.getEntranceFee() + 1000
    print(f"entrance fee is: {entrance_fee}")
    conversion_rate = fund_me.getConversionRate(entrance_fee)
    print(f"entrance fee convertet into USD in WEI format: {conversion_rate}")
    minimumUSD = 50 * 10 ** 18
    assert conversion_rate >= minimumUSD

    tx = fund_me.fund({"from": account, "value": entrance_fee})
    tx.wait(1)
    assert fund_me.addressToAmountFunded(account.address) == entrance_fee
    
    print(f"balance before withdrawal {account.balance()}")
    tx2 = fund_me.withdraw({"from": account})
    tx2.wait(1)
    print(f"balance after withdrawal {account.balance()}")
    assert fund_me.addressToAmountFunded(account.address) == 0
```

Here I add a test for the assumption that the entrance_fee is actually sufficient to clear the participation hurdle. To make this even clearer in the test, I have added two print statements that show the sums to be compared.

The above code will break at the assert and tell me something useful
```
>       assert conversion_rate >= minimumUSD
E       assert 49999999999999999140 >= 50000000000000000000
```

To be able to read the included print statements in the test, you have to start the test with the -s flag.
`(.venv) fund_me>brownie test -s`

In real life, of course, I would round the numbers generously to ensure that such a thing could not happen.

disclaimer: my entire codebase in this lesson differs a little from the repository, so you might run into bugs when directly copying it into your code. I am an absolute noob, when it comes to writing tests, I am sure, there is a much more elegant way of doing it. Use the comments to point out better ways of doing this or ask any related questions.

**Comment by cromewar** [comment]
Nice approach @n4n0b1t3, also please everyone take in consideration the base price Patrick uses as reference es for ETH to be 2000 USD dollars, if you want a more precise approach please use this links:

1. [eth-converter](https://eth-converter.com/).
2. [Eth to USD calc](https://coinmarketcap.com/converter/usd/eth/).

**Reply by n4n0b1t3** [reply]
I just noticed how often it happens, that I am reading about the same issue in the posts. I think it is easier to just link to a solution rather than explain it over and over again.

**Reply by cromewar** [reply]
Haha yes, this it's my bad habit. @PatrickAlphaC already told me I would have to point to the existing solution on another post rather than explain it over again.
I think is easier for me to just write the potential solution rather than search into the repo where that problem was posted already.
Thanks for the hint tough @n4n0b1t3

**Reply by n4n0b1t3** [reply]
That is why I write posts under the label "idea" or "general". When I filter in the search bar for "author", it recommends automatically "author:n4n0b1t3", this way I can get an overview of my posts comfortably and select one for linking.
